File: frontend/src/ai/thresholds.ts
import type { AISettings } from './types'
import type { ChemicalImportRecord } from '../imports/types'

type ChemicalThresholdSource = Pick<ChemicalImportRecord, 'name' | 'threshold'>

export function getEffectiveChemicalThreshold(chemical: ChemicalThresholdSource, settings: AISettings) {
  const override = settings.thresholds.chemicalThresholdOverrides[chemical.name]
  if (Number.isFinite(override) && override > 0) return override

  if (Number.isFinite(chemical.threshold) && chemical.threshold > 0) return chemical.threshold

  return settings.thresholds.defaultLowStockThreshold
}

File: python_backend/app/inventory/thresholds.py
"""Inventory threshold resolution helpers."""

from typing import Any


def as_number(value: Any, fallback: float = 0) -> float:
    try:
        return float(value)
    except (TypeError, ValueError):
        return fallback


def get_effective_chemical_threshold(chem: dict[str, Any], thresholds: Any | None) -> float:
    """Resolve the low-stock threshold from system settings.

    System settings are the source of truth. Per-chemical overrides are matched
    by chemical name; the imported row threshold is only a fallback for older
    deployments without saved settings.
    """
    if thresholds is not None:
        overrides = getattr(thresholds, "chemicalThresholdOverrides", {}) or {}
        override = overrides.get(str(chem.get("name") or ""))
        if override is not None:
            return as_number(override)

        default_threshold = getattr(thresholds, "defaultLowStockThreshold", None)
        if default_threshold is not None:
            return as_number(default_threshold)

    return as_number(chem.get("threshold"))
